lib/Runtime/Library/WebAssemblyMemory.cpp

```diff
@@ -139,9 +139,34 @@ WebAssemblyMemory::GrowInternal(uint32 deltaPages)
         return -1;
     }
 
-    ArrayBuffer * newBuffer = m_buffer->TransferInternal(newBytes);
-    m_buffer = newBuffer;
+    ArrayBuffer * newBuffer = nullptr;
+    JavascriptExceptionObject* caughtExceptionObject = nullptr;
+    try
+    {
+        newBuffer = m_buffer->TransferInternal(newBytes);
+    }
+    catch (const JavascriptException& err)
+    {
+        caughtExceptionObject = err.GetAndClear();
+        Assert(caughtExceptionObject);
+        //Propagate if not OOM
+        if (caughtExceptionObject != ThreadContext::GetContextForCurrentThread()->GetPendingOOMErrorObject())
+        {
+            caughtExceptionObject = caughtExceptionObject->CloneIfStaticExceptionObject(GetScriptContext());
+            JavascriptExceptionOperators::DoThrow(caughtExceptionObject, GetScriptContext());
+        }
+        GetRecycler()->ReportExternalMemoryAllocation(oldBytes); //transferInternal overestimates the amount of freed memory
```

Comment:
Looks like this happens every time an existing buffer is "resized". 
`transferInternal` calls `recycler->ReportExternalMemoryAllocation(newBufferLength - this->bufferLength)` and then `ReportExternalMemoryFailure(newBufferLength)` in case of OOM.

If OOM exception is thrown, the old buffer is left unmodified and its memory (`this->bufferLength`) also gets freed at shutdown. 

https://github.com/Microsoft/ChakraCore/pull/2137#issuecomment-267388954 provides a testcase for exposing this issue.